# Task 4.5 - Data Consistency Checks

# Exercise part 

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import os

In [2]:
# Create a path
path = r'C:\Users\pinto\OneDrive\Documenti\1-2024-Instacart Basket Analysis'

In [3]:
# Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [4]:
# Import orders.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [5]:
# Create a dataframe
df_test = pd.DataFrame()

In [6]:
#Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [7]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [8]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].map(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [9]:
# Find missing observations
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [10]:
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [11]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# Check number of rows in product dataframe
df_prods.shape

(49693, 5)

In [13]:
# Create new dataframe for non missing values
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [14]:
df_prods_clean.shape

(49677, 5)

In [15]:
# Look for duplicates in products dataframe
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [16]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
df_prods_clean.shape

(49677, 5)

In [18]:
# Create a dataframe that doesn't include the duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
df_prods_clean_no_dups.shape

(49672, 5)

In [20]:
# Export new dataframe as products_checked.csv
df_prods.to_csv(os.path.join(path, '02 Data','Prepared Data', 'products_checked.csv'))

# Task part

2. Run the df.describe() function on your df_ords dataframe. Using your new knowledge about how to interpret the output of this function, share in a markdown cell whether anything about the data looks off or should be investigated further.

In [21]:
df_ords.describe()

Unnamed: 0      order_id       user_id  order_number     order_dow  \
count  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06  3.421083e+06   
mean   1.710541e+06  1.710542e+06  1.029782e+05  1.715486e+01  2.776219e+00   
std    9.875817e+05  9.875817e+05  5.953372e+04  1.773316e+01  2.046829e+00   
min    0.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00  0.000000e+00   
25%    8.552705e+05  8.552715e+05  5.139400e+04  5.000000e+00  1.000000e+00   
50%    1.710541e+06  1.710542e+06  1.026890e+05  1.100000e+01  3.000000e+00   
75%    2.565812e+06  2.565812e+06  1.543850e+05  2.300000e+01  5.000000e+00   
max    3.421082e+06  3.421083e+06  2.062090e+05  1.000000e+02  6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

It seems that the count for all columns is the same, except for "days_since_prior order" column. This could mean there are some missing values in that column we have to investigate on.

3. Check for mixed-type data in your df_ords dataframe.

In [22]:
# Check for mixed types
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].map(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

There are no mixed data types in the df_ords dataframe.

5. Run a check for missing values in your df_ords dataframe. In a markdown cell, report your findings and propose an explanation for any missing values you find.

In [23]:
# Look for missing values in the df_ords dataframe
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
eval_set                       0
order_number                   0
order_dow                      0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

As we found out before, there are many missing values(206209) in the "days_since_prior_order" column. This could mean that some clients have placed an order for the first time.

6. Address the missing values using an appropriate method. In a markdown cell, explain why you used your method of choice.

In [24]:
# Flag missing values
df_first_order = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [25]:
df_first_order.head()

Unnamed: 0  order_id  user_id eval_set  order_number  order_dow  \
0            0   2539329        1    prior             1          2   
11          11   2168274        2    prior             1          2   
26          26   1374495        3    prior             1          1   
39          39   3343014        4    prior             1          6   
45          45   2717275        5    prior             1          3   

    order_hour_of_day  days_since_prior_order  
0                   8                     NaN  
11                 11                     NaN  
26                 14                     NaN  
39                 11                     NaN  
45                 12                     NaN

I decided to create a new dataframe containing all missing values in the "days_since_prior_order" column. I chose to do so, instead of deleting the missing values, since they hold important information; in fact, they record the number of clients making their first order.

7. Run a check for duplicate values in your df_ords data. In a markdown cell, report your findings and propose an explanation for any duplicate values you find.

In [26]:
df_dups_order = df_ords[df_ords.duplicated()]

In [27]:
df_dups_order

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, eval_set, order_number, order_dow, order_hour_of_day, days_since_prior_order]
Index: []

There are no full duplicates in the order dataframe.

9. Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder and give them appropriate, succinct names.

In [28]:
# Export new dataframe as orders_checked.csv
df_ords.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))